artemis-server/src/test/java/org/apache/activemq/artemis/tests/util/ActiveMQTestBase.java

```diff
@@ -235,29 +222,40 @@ public MBeanServer getMBeanServer() {
    // There is a verification about thread leakages. We only fail a single thread when this happens
    private static Set<Thread> alreadyFailedThread = new HashSet<>();
 
-   private LinkedList<RunnableEx> runAfter;
+   private volatile List<RunnableEx> runAfter;
 
    protected synchronized void runAfter(RunnableEx run) {
       Assert.assertNotNull(run);
       if (runAfter == null) {
-         runAfter = new LinkedList();
+         runAfter = Collections.synchronizedList(new ArrayList<>());
       }
       runAfter.add(run);
    }
 
    @After
    public void runAfter() {
-      if (runAfter != null) {
-         runAfter.forEach((r) -> {
+      List<RunnableEx> theRunafterList = runAfter;
+      if (theRunafterList != null) {
+         // I do this replacement as eventually a runAfter lambda could be adding another runAfter.
+         // this shouldn't happen very often but I saw it happening once.
+         // it's better to be protected against
```

Comment:
What did you see doing it? Is it really always invalid? If it is, then rather than silently drop it like this does, it might be better to be aware of it happening...e.g set a boolean noting the 'runAfter' things are starting to execute and then it can at least throw or log if anything calls runAfter(Runnable) to add a new task, and not add the task, rather than just allowing it to create a new collection and then ignore its contents before nulling it again as this change seems to.

Reply:
I have a case where destroyDerby is being called as runAfter();
destroyDerby is calling the getURI, which will then turn runAfter back on.


On my case I had explicitly set runAFter() to be called even though JDBC was not used.


So, it was an artificial case, but I thought It would be better to protect it since I saw it happening.


all the calls to runAfter are from single threaded. I thought when I added the volatile that it was multi-threaded.. but it's not the case.

Reply:
AH! So I resisted the temptation to comment when passing that getter that I really didnt like the method resulting in scheduling stuff.  But I totally overloooked that was this issue. lol.

That seals it, I really dont think that URI getter should be doing that, something else should be specifically opening the DB, which calls the URI getter.